packages/context/test/acceptance/extension-point.ts

```diff
@@ -0,0 +1,215 @@
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class LDAPStrategy implements AuthenticationStrategy {
+    constructor(@inject('config') public config: any) {}
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class OAuth2Strategy implements AuthenticationStrategy {
+    constructor(@inject('config') public config: any) {}
+
+    authenticate(credentials: any) {
+      return Promise.resolve(true);
+    }
+  }
+
+  class AuthenticationManager extends ExtensionPoint<AuthenticationStrategy> {
+    constructor(
+      @inject('$context') context: Context,
+      @inject('authentication.config') config: ExtensionPointConfig,
+    ) {
+      super('authentication.strategies', context, config);
+    }
+
+    async authenticate(strategy: string, credentials: any): Promise<boolean> {
+      const ext: AuthenticationStrategy = await this.getExtension(strategy);
+      return ext.authenticate(credentials);
+    }
+  }
+
+  it('lists all extensions', async () => {
+    const authManager = await ctx.get('authentication.manager');
+    const extBindings = authManager.getExtensionBindings();
+    expect(extBindings.length).to.eql(3);
+  });
+
+  it('gets an extension by name', async () => {
+    const authManager: AuthenticationManager = await ctx.get(
+      'authentication.manager',
+    );
+    const binding = authManager.getExtensionBinding('ldap');
+    expect(binding.key).to.eql('authentication.strategies.ldap');
+    expect(binding.valueConstructor).to.exactly(LDAPStrategy);
+  });
+
+  it('gets an extension instance by name', async () => {
+    const authManager: AuthenticationManager = await ctx.get(
+      'authentication.manager',
+    );
+    const ext = await authManager.getExtension('ldap');
+    expect(ext.config).to.eql({
+      url: 'ldap://localhost:1389',
+    });
+    const result = await ext.authenticate({});
+    expect(result).to.be.true();
+  });
+
+  it('delegates to an extension', async () => {
+    const authManager: AuthenticationManager = await ctx.get(
+      'authentication.manager',
+    );
+    const result = await authManager.authenticate('local', {
+      username: 'my-user',
+      password: 'my-pass',
+    });
+    expect(result).to.be.true();
+  });
+
+  function createContext() {
+    ctx = new Context();
+
+    ctx.bind('$context').to(ctx);
+    // Register the extension point
+    ctx.bind('authentication.manager').toClass(AuthenticationManager);
+    ctx.bind('authentication.config').to({
+      extensions: {
+        local: {
+          url: 'https://localhost:3000/users/login',
+        },
+        ldap: {
+          url: 'ldap://localhost:1389',
+        },
+        oauth2: {
+          clientId: 'my-client-id',
+          clientSecret: 'my-client-secret',
+          tokenInspectUrl: 'https://localhost:3000/oauth2/inspect',
+        },
+      },
+    });
+
+    // Register multiple extensions
+    ctx
```

Comment:
I'd like to see some kind of helpers to simplify the registration of extensions. Ideally, I'd want the registration code to refer to extension point by constructor (`AuthenticationHandler`) and then have everything else filled for me by the framework.

For example:

```ts
app.bindExtensionOf(AuthenticationHandler).toClass(LocalStrategy);
```

(I am using `AuthenticationHandler` instead of `AuthenticationManager` per my earlier comment.)

Reply:
+1.